### DISCUSSION TITLE
Compilation Errors on MacOS Ventura 13.3.1 with JDK 17

### DISCUSSION THREAD

**Discussion by simonjolly** [Audiveris/audiveris#646] [Q&A]
I've just installed JDK 17 on my Apple Silicon MacBook Pro running MacOS 13.3.1 and cannot get Audiveris to compile. If I clone the git repository then run:

$ ./gradlew clean build

...I always end up with the following compilation errors:

Could not initialize class org.codehaus.groovy.reflection.ReflectionCache
> Exception java.lang.NoClassDefFoundError: Could not initialize class org.codehaus.groovy.vmplugin.v7.Java7 [in thread "Daemon worker"]

Subsequently rerunning gradle gives a different error:

Could not initialize class org.codehaus.groovy.runtime.InvokerHelper
> Exception java.lang.NoClassDefFoundError: Could not initialize class org.codehaus.groovy.reflection.ReflectionCache [in thread "Daemon worker"]

Any suggestions on what's going wrong and how to get Audiveris to compile?

**Comment by hbitteur** [comment]
@simonjolly 
I have no access to a Mac and cannot really help you. I will forward your question to @maximumspatium for a better answer.

Just an idea: instead of default master branch (meant to be used with Java 11), please select the development branch (meant to be used with Java 17, and providing more recent features).

To do so, use commands:
> git checkout development
> ./gradlew clean build

Please try it and keep us informed of the results.

**Comment by simonjolly** [comment]
No issue with checking out the development branch, but the build still fails, this time with a different error:

FAILURE: Build failed with an exception.

* Where:
Settings file '/Users/jolly/bin/audiveris/settings.gradle'

* What went wrong:
Could not compile settings file '/Users/jolly/bin/audiveris/settings.gradle'.
> startup failed:
  General error during conversion: Unsupported class file major version 63

I note that this time Gradle 7.3 has been downloaded rather than 5.6.4 for the master branch. Here's the first few lines of the exception thrown:

  java.lang.IllegalArgumentException: Unsupported class file major version 63
        at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:199)
        at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:180)
        at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:166)
        at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:287)
        at org.codehaus.groovy.ast.decompiled.AsmDecompiler.parseClass(AsmDecompiler.java:81)
        at org.codehaus.groovy.control.ClassNodeResolver.findDecompiled(ClassNodeResolver.java:251)
        at org.codehaus.groovy.control.ClassNodeResolver.tryAsLoaderClassOrScript(ClassNodeResolver.java:189)
        at org.codehaus.groovy.control.ClassNodeResolver.findClassNode(ClassNodeResolver.java:169)
        at org.codehaus.groovy.control.ClassNodeResolver.resolveName(ClassNodeResolver.java:125)
        at org.codehaus.groovy.ast.decompiled.AsmReferenceResolver.resolveClassNullable(AsmReferenceResolver.java:57)
        at org.codehaus.groovy.ast.decompiled.AsmReferenceResolver.resolveClass(AsmReferenceResolver.java:44)
        at org.codehaus.groovy.ast.decompiled.TypeSignatureParser.visitEnd(TypeSignatureParser.java:110)
        at groovyjarjarasm.asm.signature.SignatureReader.parseType(SignatureReader.java:206)
        at groovyjarjarasm.asm.signature.SignatureReader.accept(SignatureReader.java:124)

**Reply by hbitteur** [reply]
> Unsupported class file major version 63

This sounds like a mismatch with Java versions.

Are you sure you are using Java 17?
Please check with:
> java -version

Also please check JAVA_HOME environment variable with:
> echo $JAVA_HOME

**Reply by hbitteur** [reply]
Could you post the content of file /Users/jolly/bin/audiveris/settings.gradle
Just in case...

**Reply by simonjolly** [reply]
Here you go:

$ java -version
java version "19.0.2" 2023-01-17
Java(TM) SE Runtime Environment (build 19.0.2+7-44)
Java HotSpot(TM) 64-Bit Server VM (build 19.0.2+7-44, mixed mode, sharing)

$ more settings.gradle 
rootProject.name = 'Audiveris'
include 'schemas'

$ echo $JAVA_HOME



(Yes, it was blank...)

**Comment by hbitteur** [comment]
Don't use Java 19 but Java 17! According to your first message, you had installed Java 17, not Java 19. How come?

Also set JAVA_HOME to root folder of Java 17

**Comment by simonjolly** [comment]
OKay, I managed to make a little more progress. I'm genuinely nonplussed about the Java version: I simply downloaded the JDK 17 disk image from the Oracle download page:

https://www.oracle.com/java/technologies/downloads/#java17

I can't even see that JDK 19 is available...

Anyway, having removed Java, redownloaded it and reinstalled it — as well as setting JAVA_HOME=/usr/libexec/java_home — gradle now seems to be running correctly. I now get the following Java version:

$ java -version
java version "17.0.7" 2023-04-18 LTS
Java(TM) SE Runtime Environment (build 17.0.7+8-LTS-224)
Java HotSpot(TM) 64-Bit Server VM (build 17.0.7+8-LTS-224, mixed mode, sharing)

Compiling required checking out the development branch, but Gradle now fails with the following exception:

> Task :startScripts FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':startScripts'.
> Error while evaluating property 'relativeClasspath' of task ':startScripts'
   > Could not resolve all files for configuration ':runtimeClasspath'.
      > Could not find tesseract-4.1.1-1.5.6-macosx-arm64.jar (org.bytedeco:tesseract:4.1.1-1.5.6).
        Searched in the following locations:
            https://repo.maven.apache.org/maven2/org/bytedeco/tesseract/4.1.1-1.5.6/tesseract-4.1.1-1.5.6-macosx-arm64.jar
      > Could not find leptonica-1.81.1-1.5.6-macosx-arm64.jar (org.bytedeco:leptonica:1.81.1-1.5.6).
        Searched in the following locations:
            https://repo.maven.apache.org/maven2/org/bytedeco/leptonica/1.81.1-1.5.6/leptonica-1.81.1-1.5.6-macosx-arm64.jar

I had assumed this is because it's looking for Tesseract data files. I've cloned both the Tessaract and Tesseract "tessdata" repository and put a soft link inside the audiveris directory, but it hasn't made any difference. I'm slightly nonplussed about it throwing exceptions about missing JAR files though.

If you can let me know which specific files are needed from Tesseract, where they can be downloaded from and where they should be copied to in order for Audiveris to find them, that would be great.

So two steps forward, one step...

**Reply by simonjolly** [reply]
Just as an addendum, I also went looking in the byte deco repository that is linked in the exception message. As the message notes, there is no arm64 JAR file for that version of Tesseract. The only release that has a version compiled for Apple Silicon Macs (ARM64) is 5.2.0:

https://repo.maven.apache.org/maven2/org/bytedeco/tesseract/5.2.0-1.5.8/

Downloading this and putting it into the Audiveris root directory didn't work, even after I renamed it to the 4.1.1 filename Gradle was looking for.
